edb/server/protocol/protocol.pyx

```diff
@@ -75,25 +79,50 @@ cdef class HttpProtocol:
         self.transport = None
         self.unprocessed = None
 
+    def eof_received(self):
+        pass
+
     def data_received(self, data):
         if self.first_data_call:
             self.first_data_call = False
 
+            is_ssl = True
+            is_binary = False
+            try:
+                outgoing = ssl.MemoryBIO()
+                incoming = ssl.MemoryBIO()
+                incoming.write(data)
+                sslobj = self.server._sslctx.wrap_bio(
+                    incoming, outgoing, server_side=True
+                )
+                sslobj.do_handshake()
+            except ssl.SSLWantReadError:
+                pass
+            except Exception:
+                is_ssl = False
+                if not self.server._tls_compat:
+                    if data[0:2] == b'V\x00':
+                        # This is, most likely, our binary protocol,
+                        # as its first message kind is `V`.
+                        self.loop.create_task(self._return_binary_error())
+                    else:
+                        self.response_hsts = True
+
+            if is_ssl:
+                self.loop.create_task(self._forward_first_data(data))
+                self.loop.create_task(self._start_tls())
+                return
+
             if data[0:2] == b'V\x00':
                 # This is, most likely, our binary protocol,
                 # as its first message kind is `V`.
                 #
                 # Switch protocols now (for compatibility).
-                binproto = binary.EdgeConnection(
-                    self.server, self.external_auth)
-                self.transport.set_protocol(binproto)
-                binproto.connection_made(self.transport)
-                binproto.data_received(data)
+                self._switch_to_binary_protocol(data)
```

Comment:
Only when `if self.server._tls_compat`?

Reply:
My bad ... this is actually needed by `_return_binary_error()` - I'll make it more readable.